cloudify/decorators.py

```diff
@@ -40,4 +45,87 @@ def workflow(func=None, system_wide=False, resumable=False, **kwargs):
+        to. The default is ['install']. This means we scale will
+        execute in parallel.
+    :param states: List of states that the preceding
+        node instance may be in before executing the current node instance.
+        The default is ['started']. We will only continue to next instance
+        after the preceding instance is in "started" state.
+    :raises: CloudifySerializationRetry.
+    :return: decorator wrapper
+    """
+
+    workflows = workflows if isinstance(
+        workflows, list) else ['install']
+    states = states if isinstance(
+        states, list) else ['started']
+    rest_client = get_rest_client()
+
+    def outer_wrapper(func):
+        def inner_wrapper(*args, **kwargs):
+            _ctx = kwargs.get('ctx')
+            if _ctx.workflow_id in workflows:
+                # Split the instances at the index into
+                # "priority_instances" and "no_priority"
+                # node instances. "priority_instances means that these
+                # instances' index is a smaller number than the
+                # current node instances's. They should be executed first.
+                # "no_priority" means that it comes after.
+                priority_instances, no_priority = \
+                    split_instances_at_index(rest_client)
+                # Finished instances are those that are already in the
+                # state where they are not considered blockers of our
+                # current node instance.
+                non_blocking_instances = [
+                    i for i in priority_instances + no_priority
+                    if i.state in states]
+                try:
+                    # preceding_instance is the instance directly preceding
+                    # the current node instance according to index.
+                    preceding_instance = priority_instances[-1]
+                except IndexError:
+                    # If there are no priority_instances instances, we can just
+                    # run the operation.
+                    pass
+                else:
+                    # We want to wait if the following conditions exist:
+                    # 1) revert_to_parallel_threshold is 0 and preceding
+                    #    node instance (preceding_instance) is not ready.
+                    #    This means - always try to execute pure serial.
+                    # 2) revert_to_parallel_threshold is unequal to total
+                    #    finished instances. This means that if we have
+                    #    exceeded the revert_to_parallel_threshold threshold,
+                    #    then we would return to parallel execution.
+                    if (revert_to_parallel_threshold == 0 or
+                        revert_to_parallel_threshold != len(
+                                non_blocking_instances)) and \
+                            preceding_instance and \
+                            preceding_instance.state not in states:
```

Comment:
oh, i understand - this actually requires the sort, because we want to find the very previous instance to check its state

Reply:
then maybe write this more directly, without splitting, and just find the preceding instance using a forloop or otherwise
(eg. `preceding_instance = next((ni for ni in instances if ni.index == current_ni.index - 1), None)` or `preceding = None;  for ni in instances:  if ni.index == current_ni.index - 1: preceding = ni; break`)